File: MyScript/edk/java/examples/ink/UsingTags.java
package ink;

import java.io.File;
import java.io.IOException;

import certificates.MyCertificate;

import com.myscript.engine.Engine;
import com.myscript.engine.EngineObject;
import com.myscript.geometry.Rectangle;
import com.myscript.ink.Ink;
import com.myscript.ink.InkSegmentMatchMode;
import com.myscript.ink.InkSelection;
import com.myscript.ink.InkTagIterator;
import com.myscript.ink.InkTagQuery;
import com.myscript.ink.SelectionModifier;
import com.myscript.json.Json;
import com.myscript.json.JsonObjectIterator;
import common.CFG;

/**
 * The <code>UsingTags/code> example code shows you how to use tags within ink.
 */
public class UsingTags
{
  /**
   * Path to the .jink file
   */
  private static final String JINK_FILE = CFG.DATA_DIR + "/flower.jink";

  /**
   * Main entry point for the <code>UsingTags</code> example program.
   * 
   * @param args command line arguments, ignored.
   * @throws IOException
   */
  public static final void main(String[] args) throws IOException
  {
    if (JINK_FILE == null || JINK_FILE.length() == 0)
    {
      System.err.println("/!\\ please choose a valid filename for jink file. You can choose document filename by editing JINK_FILE static member variable of UsingTags class");
      System.exit(-1);
    }

    System.out.println("1) creating a new Engine initialized with your certificate");
    final Engine engine = Engine.create(MyCertificate.getBytes());
    System.out.println(" . Engine created successfully");
    System.out.println();

    System.out.println("2) loading an existing .jink file");
    final Ink ink = (Ink)Ink.load(engine, JINK_FILE);    
    System.out.println(" . " + new File(JINK_FILE).getCanonicalPath() + " loaded successfully");
    System.out.println();

    System.out.println("3) inspecting the content of the loaded Ink");
    final InkTagQuery tagQuery = new InkTagQuery();
    tagQuery.mode = InkSegmentMatchMode.MATCH_INCLUDE;
    tagQuery.intervals = null;
    int tagCount = ink.tagLookupCount(tagQuery);
    System.out.println(" . the Ink contains " + tagCount + " tags");
    System.out.println();

    System.out.println("4) selecting a portion of Ink");
    final InkSelection selection = InkSelection.create(ink);
    // select a portion of ink
    final Rectangle rectangle = new Rectangle(230, 300, 50, 50);
    selection.selectRectangle(rectangle, SelectionModifier.SET);
    // check that we selected something
    System.out.println(" . selection created with " + selection.getStrokeCount() + " strokes");
    System.out.println();

    System.out.println("5) adjusting the selection to tag boundaries");
    selection.adjustToTagBoundaries("chinese_word", 0.3f);
    System.out.println(" . the adjusted selection now contains " + selection.getStrokeCount() + " strokes");
    System.out.println();

    System.out.println("6) inspecting the tags intersecting with this selection");
    tagQuery.intervals = selection;
    tagQuery.mode = InkSegmentMatchMode.MATCH_INTERSECT;
    tagCount = ink.tagLookupCount(tagQuery);
    System.out.println(" . the adjusted selection intersects with " + tagCount + " tags:");
    InkTagIterator tagIterator = ink.tagLookup(tagQuery);
    while(false == tagIterator.isAtEnd())
    {
      System.out.println("     tag id = " + tagIterator.getId() + " , tag name = \"" + tagIterator.getName() + "\"");
      tagIterator.next();
    }
    tagIterator.dispose();
    System.out.println();

    tagQuery.intervals = null;
    tagQuery.id = new Long(1);
    System.out.println("7) selecting the tag with id " + tagQuery.id);
    tagIterator = ink.tagLookup(tagQuery);
    EngineObject tagData = null;
    if (!tagIterator.isAtEnd())
    {
      System.out.println(" . the name of this tag is \"" + tagIterator.getName() + "\"");
      tagData = tagIterator.getData();
      if (tagData != null)
      {
        if (tagData instanceof Json)
        {
          System.out.println(" . the tag has associated data of type JSON");
        }
      }
      tagIterator.next();
    }
    tagIterator.dispose();
    System.out.println();

    System.out.println("8) inspecting JSON data associated with tag " + tagQuery.id);
    final Json jsonObject = (Json)(tagData);
    final JsonObjectIterator jsonIterator = jsonObject.getObjectEntries();
    System.out.println(" . the JSON data is a JSON object which contains the following keys:");
    while(false == jsonIterator.isAtEnd())
    {
      System.out.println("     key = \"" + jsonIterator.getKey() + "\"");
      jsonIterator.next();
    }
    jsonIterator.dispose();
    tagData.dispose();
    System.out.println();

    System.out.println("9) erasing the selection");
    ink.erase(selection);
    tagQuery.id = new Long(-1);
    tagCount = ink.tagLookupCount(tagQuery);
    System.out.println(" . the Ink contains " + tagCount + " tags after a selection removal");
    System.out.println();
    
    tagQuery.id = new Long(1);
    System.out.println("10) erasing tag with id " + tagQuery.id);
    tagIterator = ink.tagLookup(tagQuery);
    if (!tagIterator.isAtEnd())
    {
      tagIterator.remove();
      tagIterator.next();
      tagIterator.dispose();
      tagQuery.id = new Long(-1);
      tagCount = ink.tagLookupCount(tagQuery);
      System.out.println(" . the Ink contains " + tagCount + " tags after tag removal");
    }
    System.out.println();

    // do not forget the 1st rule: you create it, you dispose of it!
    System.out.println("11) cleaning-up: disposing of objects and disposing of the Engine");
    selection.dispose();
    ink.dispose();
    engine.dispose();
    System.out.println(" . clean-up done successfully");
    System.out.println();

    System.out.println("end of example:");
    System.out.println(" you just learnt how to use tags within ink");
  }

} // UsingTags

File: MyScript/edk/c/examples/ink/UsingTags.c
/**
 * The `UseTags` example code shows you how to use tags within ink.
 */


// -- configuration ------------------------------------------------------------

// common configuration parameters
#include "../common/config.h"

// include the MyScript Engine API header(s)
#include <c/MyScriptEngine.h>
#include <c/MyScriptInk.h>
#include <c/MyScriptJSON.h>

// include the certificate
#include "../certificates/MyCertificate.h"


// -- example body -------------------------------------------------------------

#include <stdio.h>
#include <stdlib.h>

// global engine (for the sake of readability)
voEngine engine = NULL;

// common utility functions
#include "../common/common.h"

// path to the .jink file
#define JINK_FILE DATA_DIR "/flower.jink"

#define BUFFER_SIZE 200

// main function
int main(void)
{
  voInk ink;
  int tagCount;
  voRectangle rectangle;
  voInkSelection selection;
  int strokeCount;
  float overlapRatio;
  voInkTagQuery tagQuery = voInkTagQuery_EMPTY_INITIALIZER;
  voInkTagIterator tagIterator;
  voString string_chinese = voString_INITIALIZER("chinese_word");
  int64_t tagId;
  char stringBuffer[BUFFER_SIZE];
  voString string = { stringBuffer, BUFFER_SIZE };
  voEngineObject tagData;
  voJsonObjectIterator jsonIterator;

  // -- initialization ---------------------------------------------------------
  if (!loadMyScriptEngine(MYSCRIPT_ENGINE_LIBRARY_DIR, NULL))
    return (fprintf(stderr, "failed to load the MyScriptEngine library\n"), EXIT_FAILURE);

  // -- example code -----------------------------------------------------------
  printf("1) creating a new Engine initialized with your certificate\n");
  engine = voCreateEngine(VO_MSE_VER, &myCertificate, NULL);
  verify(engine != NULL);
  printf(" . Engine created successfully\n");
  printf("\n");

  printf("2) loading an existing .jink file\n");
  ink = loadEngineObject(JINK_FILE);
  verify(ink != NULL);
  printf(" . ");
  printPath(stdout, JINK_FILE);
  printf(" loaded successfully\n");
  printf("\n");

  printf("3) inspecting the content of the loaded Ink\n");
  tagCount = VO_IFACE(engine, IInk)->tagLookupCount(engine, ink, &tagQuery);
  verify(tagCount > -1);
  printf(" . the Ink contains %d tags\n", tagCount);
  printf("\n");

  printf("4) selecting a portion of Ink\n");
  selection = voCreateInkSelection(engine, ink);
  verify(selection != NULL);
  // select a portion of ink
  rectangle.x = 230;
  rectangle.y = 300;
  rectangle.width = 50;
  rectangle.height = 50;
  verify(VO_IFACE(engine, ISelection)->selectRectangle(engine, selection, &rectangle, VO_SELECT_SET));
  strokeCount = VO_IFACE(engine, IInkStrokeSet)->getStrokeCount(engine, selection);
  printf(" . selection created with %d strokes\n", strokeCount);
  printf("\n");

  printf("5) adjusting the selection to tag boundaries\n");
  overlapRatio = 0.3f;
  verify(VO_IFACE(engine, IInkSelection)->adjustToTagBoundaries(engine, selection, NULL, &string_chinese, overlapRatio));
  strokeCount = VO_IFACE(engine, IInkStrokeSet)->getStrokeCount(engine, selection);
  printf(" . the adjusted selection now contains %d strokes\n", strokeCount);
  printf("\n");

  printf("6) inspecting the tags intersecting with this selection\n");
  tagQuery.intervals = selection;
  tagQuery.mode = VO_INKINTERVALS_MATCH_INTERSECT;
  tagCount = VO_IFACE(engine, IInk)->tagLookupCount(engine, ink, &tagQuery);
  printf(" . the adjusted selection intersects with %d tags:\n", tagCount);
  tagIterator = VO_IFACE(engine, IInk)->tagLookup(engine, ink, &tagQuery);
  verify(tagIterator != NULL);
  while(VO_NO == VO_IFACE(engine, IInkTagIterator)->isAtEnd(engine, tagIterator))
  {
    tagId = VO_IFACE(engine, IInkTagIterator)->getId(engine, tagIterator);
    string.byteCount = BUFFER_SIZE;
    verify(VO_IFACE(engine, IInkTagIterator)->getName(engine, tagIterator, NULL, &string));
    verify(string.byteCount < BUFFER_SIZE);
    string.bytes[string.byteCount] = '\0';
    printf("     tag id = %lld, tag name = \"%s\"\n", tagId, string.bytes);
    verify(VO_IFACE(engine, IInkTagIterator)->next(engine, tagIterator));
  }
  verify(voReleaseObject(engine, tagIterator));
  printf("\n");

  tagQuery.intervals = NULL;
  tagQuery.id = 1;
  printf("7) selecting the tag with id %lld\n", tagQuery.id);
  tagIterator = VO_IFACE(engine, IInk)->tagLookup(engine, ink, &tagQuery);
  verify(tagIterator != NULL);
  verify(VO_NO == VO_IFACE(engine, IInkTagIterator)->isAtEnd(engine, tagIterator));
  string.byteCount = BUFFER_SIZE;
  verify(VO_IFACE(engine, IInkTagIterator)->getName(engine, tagIterator, NULL, &string));
  verify(string.byteCount < BUFFER_SIZE);
  string.bytes[string.byteCount] = '\0';
  printf(" . the name of this tag is \"%s\"\n", string.bytes);
  tagData = VO_IFACE(engine, IInkTagIterator)->getData(engine, tagIterator);
  verify(tagData != NULL);
  verify(VO_Json == voGetType(engine, tagData));
  printf(" . the tag has associated data of type JSON\n");
  verify(VO_IFACE(engine, IInkTagIterator)->next(engine, tagIterator));
  verify(VO_YES == VO_IFACE(engine, IInkTagIterator)->isAtEnd(engine, tagIterator));
  verify(voReleaseObject(engine, tagIterator));
  printf("\n");

  printf("8) inspecting JSON data associated with tag %lld\n", tagQuery.id);
  jsonIterator = VO_IFACE(engine, IJson)->getObjectEntries(engine, tagData);
  verify(jsonIterator != NULL);
  printf(" . the JSON data is a JSON object which contains the following keys:\n");
  while(VO_NO == VO_IFACE(engine, IJsonObjectIterator)->isAtEnd(engine, jsonIterator))
  {
    string.byteCount = BUFFER_SIZE;
    verify(VO_IFACE(engine, IJsonObjectIterator)->getKey(engine, jsonIterator, NULL, &string));
    verify(string.byteCount < BUFFER_SIZE);
    string.bytes[string.byteCount] = '\0';
    printf("     key = \"%s\"\n", string.bytes);
    verify(VO_IFACE(engine, IJsonObjectIterator)->next(engine, jsonIterator));
  }
  verify(voReleaseObject(engine, jsonIterator));
  verify(voReleaseObject(engine, tagData));
  printf("\n");

  printf("9) erasing the selection\n");
  verify(VO_IFACE(engine, IInk)->erase(engine, ink, selection));
  tagQuery.id = -1;
  tagCount = VO_IFACE(engine, IInk)->tagLookupCount(engine, ink, &tagQuery);
  verify(tagCount > -1);
  printf(" . the Ink contains %d tags after a selection removal\n", tagCount);
  printf("\n");

  tagQuery.id = 1;
  printf("10) erasing tag with id %lld\n", tagQuery.id);
  tagIterator = VO_IFACE(engine, IInk)->tagLookup(engine, ink, &tagQuery);
  verify(tagIterator != NULL);
  verify(VO_NO == VO_IFACE(engine, IInkTagIterator)->isAtEnd(engine, tagIterator));
  verify(VO_IFACE(engine, IInkTagIterator)->remove(engine, tagIterator));
  verify(VO_IFACE(engine, IInkTagIterator)->next(engine, tagIterator));
  verify(VO_YES == VO_IFACE(engine, IInkTagIterator)->isAtEnd(engine, tagIterator));
  verify(voReleaseObject(engine, tagIterator));
  tagQuery.id = -1;
  tagCount = VO_IFACE(engine, IInk)->tagLookupCount(engine, ink, &tagQuery);
  verify(tagCount > -1);
  printf(" . the Ink contains %d tags after tag removal\n", tagCount);
  printf("\n");

  printf("11) cleaning-up: releasing all created objects and destroying the Engine\n");
  verify(voReleaseObject(engine, selection));
  verify(voReleaseObject(engine, ink));
  verify(voDestroyEngine(engine));
  printf(" . clean-up done successfully\n");
  printf("\n");

  printf("end of example:\n");
  printf(" you just learnt how to use tags within ink\n");

  // -- finalization -----------------------------------------------------------
  unloadMyScriptEngine();

  return EXIT_SUCCESS;
}
